src/Principal.mo

```diff
@@ -254,5 +291,932 @@ module {
+  let K10 : Nat32 = 0x243185be;
+  let K11 : Nat32 = 0x550c7dc3;
+  let K12 : Nat32 = 0x72be5d74;
+  let K13 : Nat32 = 0x80deb1fe;
+  let K14 : Nat32 = 0x9bdc06a7;
+  let K15 : Nat32 = 0xc19bf174;
+  let K16 : Nat32 = 0xe49b69c1;
+  let K17 : Nat32 = 0xefbe4786;
+  let K18 : Nat32 = 0x0fc19dc6;
+  let K19 : Nat32 = 0x240ca1cc;
+  let K20 : Nat32 = 0x2de92c6f;
+  let K21 : Nat32 = 0x4a7484aa;
+  let K22 : Nat32 = 0x5cb0a9dc;
+  let K23 : Nat32 = 0x76f988da;
+  let K24 : Nat32 = 0x983e5152;
+  let K25 : Nat32 = 0xa831c66d;
+  let K26 : Nat32 = 0xb00327c8;
+  let K27 : Nat32 = 0xbf597fc7;
+  let K28 : Nat32 = 0xc6e00bf3;
+  let K29 : Nat32 = 0xd5a79147;
+  let K30 : Nat32 = 0x06ca6351;
+  let K31 : Nat32 = 0x14292967;
+  let K32 : Nat32 = 0x27b70a85;
+  let K33 : Nat32 = 0x2e1b2138;
+  let K34 : Nat32 = 0x4d2c6dfc;
+  let K35 : Nat32 = 0x53380d13;
+  let K36 : Nat32 = 0x650a7354;
+  let K37 : Nat32 = 0x766a0abb;
+  let K38 : Nat32 = 0x81c2c92e;
+  let K39 : Nat32 = 0x92722c85;
+  let K40 : Nat32 = 0xa2bfe8a1;
+  let K41 : Nat32 = 0xa81a664b;
+  let K42 : Nat32 = 0xc24b8b70;
+  let K43 : Nat32 = 0xc76c51a3;
+  let K44 : Nat32 = 0xd192e819;
+  let K45 : Nat32 = 0xd6990624;
+  let K46 : Nat32 = 0xf40e3585;
+  let K47 : Nat32 = 0x106aa070;
+  let K48 : Nat32 = 0x19a4c116;
+  let K49 : Nat32 = 0x1e376c08;
+  let K50 : Nat32 = 0x2748774c;
+  let K51 : Nat32 = 0x34b0bcb5;
+  let K52 : Nat32 = 0x391c0cb3;
+  let K53 : Nat32 = 0x4ed8aa4a;
+  let K54 : Nat32 = 0x5b9cca4f;
+  let K55 : Nat32 = 0x682e6ff3;
+  let K56 : Nat32 = 0x748f82ee;
+  let K57 : Nat32 = 0x78a5636f;
+  let K58 : Nat32 = 0x84c87814;
+  let K59 : Nat32 = 0x8cc70208;
+  let K60 : Nat32 = 0x90befffa;
+  let K61 : Nat32 = 0xa4506ceb;
+  let K62 : Nat32 = 0xbef9a3f7;
+  let K63 : Nat32 = 0xc67178f2;
+
+  let ivs : [[Nat32]] = [
+    [
+      // 224
+      0xc1059ed8,
+      0x367cd507,
+      0x3070dd17,
+      0xf70e5939,
+      0xffc00b31,
+      0x68581511,
+      0x64f98fa7,
+      0xbefa4fa4
+    ],
+    [
+      // 256
+      0x6a09e667,
+      0xbb67ae85,
+      0x3c6ef372,
+      0xa54ff53a,
+      0x510e527f,
+      0x9b05688c,
+      0x1f83d9ab,
+      0x5be0cd19
+    ]
+  ];
+
+  let rot = Nat32.bitrotRight;
+
+  class SHA224() {
+    let (sum_bytes, iv) = (28, 0);
+
+    var s0 : Nat32 = 0;
+    var s1 : Nat32 = 0;
+    var s2 : Nat32 = 0;
+    var s3 : Nat32 = 0;
+    var s4 : Nat32 = 0;
+    var s5 : Nat32 = 0;
+    var s6 : Nat32 = 0;
+    var s7 : Nat32 = 0;
+
+    let msg : [var Nat32] = Array.init<Nat32>(16, 0);
+    let digest = Array.init<Nat8>(sum_bytes, 0);
+    var word : Nat32 = 0;
+
+    var i_msg : Nat8 = 0;
+    var i_byte : Nat8 = 4;
+    var i_block : Nat64 = 0;
+
+    public func reset() {
+      i_msg := 0;
+      i_byte := 4;
+      i_block := 0;
+      s0 := ivs[iv][0];
+      s1 := ivs[iv][1];
+      s2 := ivs[iv][2];
+      s3 := ivs[iv][3];
+      s4 := ivs[iv][4];
+      s5 := ivs[iv][5];
+      s6 := ivs[iv][6];
+      s7 := ivs[iv][7]
```

Comment:
Just for completeness, commenting here on an already merged PR, and too minor to be worth opening this again.

Since we are only using Sha224 I would have hard-codes the values here and remove the `ivs` array entirely since it also includes IVs for Sha256 which is confusing.